EIP-4341: Ordered NFT Batch Standard (#4341)
ethereum/EIPs: EIPS/eip-4341.md

---
eip: 4341

> title: Multi Ordered NFT Standard

Reviewer: ```suggestion
title: Ordered NFT Batch Standard
```
Just a suggestion, but I find this title much more descriptive.

> description: The ordering information of multiple NFTs is retained and managed.

Reviewer: ```suggestion
description: The ordering information of multiple NFTs is retained and managed

```

author: Simon Tian (@simontianx)
discussions-to: https://github.com/ethereum/EIPs/issues/3782
status: Draft

> type: Standards Track,

Reviewer: ```suggestion
type: Standards Track
```

---
eip: 4341
title: Multi Ordered NFT Standard
description: The ordering information of multiple NFTs is retained and managed

> 

Reviewer: ```suggestion
```

---
eip: 4341
title: Unicoded NFT Standard
description: For on-chain logographic languages and emojis
author: Simon Tian (@simontianx)
discussions-to: https://github.com/ethereum/EIPs/issues/3782
status: Draft
type: Standards Track,
category: ERC
created: 2021-10-01

> replaces (*optional): <EIP number(s)>

Reviewer: ```suggestion
```

---

## Abstract
A vanilla non-fungible token standard proposed by [EIP-3754](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-3754.md)
combined with the established **[unicode](https://home.unicode.org/)** system
can create an on-chain logographic system for representing language characters
that are **[logographs](https://en.wikipedia.org/wiki/Logogram)**. Examples
include _hanzi_ in Mandarin, _kanji_ in Japanese, _hanja_ in Korean, and etc.

## Motivation
A logograph is a written character that represents a word or morpheme, whereas
a unicode is an information technology standard for the consistent encoding,

> representation, and handling of texts. If the tokenId of an NFT is specified as

Reviewer: ```suggestion
representation, and handling of texts. If the `tokenId` of an NFT is specified as

```

This standard aims to combine the two and introduce the necessary functionalities
for batch transfer and representation of logographic characters. As a result, an
on-chain language system can be built for logograpic languages.

For non-logographic languages such as English, although the same standard can
be applied, the marginal utility is minimal to be practically useful. However,
for logographic languages, since a rich amount of meanings can be transmitted in
just a few characters, this standard is therefore particularly valuable. Emojis
are similar with logographs and can be included as well.

A motivating example is instead of sending the two Chinese characters of the
Great Wall `长城`, two NFTs with IDs `#38271` and `#22478` respectively can be
transferred in a batch. The two IDs are corresponding to the decimal unicode of
the two characters. The receiving end decodes the IDs and retrieves the original
characters. A key point is the ordering information matters in this scenario
since the tuples `(38271, 22478)` and `(22478, 38271)` can be decoded as
`长城` and `城长`, respectively, and both are legitimate words in the Chinese

> language. This illustrates the key difference between this standard and ERC1155.

Reviewer: ```suggestion
language. This illustrates the key difference between this standard and [ERC-1155](./eip-1155.md).

```

 */
interface ERC4341 /* is ERC165 */ {
    event Transfer(address indexed _operator, address indexed _from, address indexed _to, uint256 _id, uint256 _value);

    event TransferBatch(address indexed _operator, address indexed _from, address indexed _to, uint256[] _ids, uint256[] _values);

    event ApprovalForAll(address indexed _owner, address indexed _operator, bool _approved);

    function safeTransferFrom(address _from, address _to, uint256 _id, uint256 _value, bytes calldata _data) external;

    function safeBatchTransferFrom(address _from, address _to, uint256[] calldata _ids, uint256[] calldata _values, bytes calldata _data) external;

    function balanceOf(address _owner, uint256 _id) external view returns (uint256);

    function balanceOfBatch(address[] calldata _owners, uint256[] calldata _ids) external view returns (uint256[] memory);

    function retrievePhrase(address _owner, uint256 _phraseId) external view returns (uint256[] memory);

    function setApprovalForAll(address _operator, bool _approved) external;

    function isApprovedForAll(address _owner, address _operator) external view returns (bool);
}
```

## Rationale

> If we think of [ERC1155](./eip-1155.md) together with [ERC721](./eip-721.md) being

Reviewer: ```suggestion
If we think of [ERC-1155](./eip-1155.md) together with [ERC-721](./eip-721.md) being
```

in the track of representing crypto-assets with tokens, then [EIP-3754](./eip-3754.md)

> and this standard EIP-4341 are in the track of representing the utility sides of NFTs.

Reviewer: ```suggestion
and this standard are in the track of representing the utility sides of NFTs.

```

### Ordering information

> Compared with ERC1155, this standard is particularly useful in logographic

Reviewer: ```suggestion
Compared with ERC-1155, this standard is particularly useful in logographic

```

> representations. Therefore, the ordering information in a batch of tokenIds is

Reviewer: ```suggestion
representations. Therefore, the ordering information in a batch of `tokenId`s is

```

> maintained whereas such information is completed discarded in ERC1155. It is not

Reviewer: ```suggestion
maintained whereas such information is completed discarded in ERC-1155. It is not

```

> a problem in ERC1155 since no valuable information is encoded in the order of

Reviewer: ```suggestion
a problem in ERC-1155 since no valuable information is encoded in the order of

```

> tokenIds, however, in this standard, as illustrated above, it is crucial in

Reviewer: ```suggestion
`tokenId`s, however, in this standard, as illustrated above, it is crucial in

```

decoding the unicodes of a sequence of logographs in the correct order.

### `URI`

> To further distinguish this standard from ERC1155, `URI`s should be excluded in

Reviewer: ```suggestion
To further distinguish this standard from ERC-1155, `URI`s should be excluded in

```
